src/libs/tmdk/tests/deploy.test.ts

```diff
@@ -3,24 +3,246 @@
-import { mockExit } from "./test-utils";
+import {
+  workspaceId,
+  s3BucketArn,
+  s3ContentLocationBase,
+  workspaceBucket,
+} from "./test-utils";
+import { mockClient } from "aws-sdk-client-mock";
+import {
+  CreateComponentTypeCommand,
+  CreateEntityCommand,
+  CreateSceneCommand,
+  GetComponentTypeCommand,
+  GetEntityCommand,
+  GetWorkspaceCommand,
+  IoTTwinMakerClient,
+  ResourceNotFoundException,
+} from "@aws-sdk/client-iottwinmaker";
+import {
+  emptyTmdk,
+  componentType1Definition,
+  scene1,
+  entity1Definition,
+} from "./test-constants";
+import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";
+import * as fsPromises from "fs/promises";
+const fs = require("fs");
 
-jest.mock("../src/lib/aws-clients");
+jest.mock("fs/promises", () => {
+  return { readFile: jest.fn() };
+});
+const twinmakerMock = mockClient(IoTTwinMakerClient);
+const s3Mock = mockClient(S3Client);
+const fakeTmdkDir = "/tmp/deploy-unit-tests";
+
+beforeEach(() => {
+  twinmakerMock.reset();
+  s3Mock.reset();
+  jest.resetAllMocks();
+});
+
+test("throws error when given tmdk project that does not exist", async () => {
+  const argv2 = {
+    _: ["deploy"],
+    $0: "tmdk_local",
+    region: "us-east-1",
+    "workspace-id": "irrelevant",
+    dir: "i-do-not-exist",
+  } as Arguments<Options>;
+  await expect(handler(argv2)).rejects.toThrow(
+    Error("TDMK.json does not exist. Please run tmdk init first.")
+  );
+});
+
+test("throws error when given workspace that does not exist", async () => {
+  fs.existsSync = jest.fn(() => true );
+  jest.spyOn(fs, "readFileSync").mockReturnValue(JSON.stringify({}, null, 4));
+  twinmakerMock
+    .on(GetWorkspaceCommand)
+    .rejects(new ResourceNotFoundException({ $metadata: {}, message: "" }));
+
+  const argv2 = {
+    _: ["init"],
+    $0: "tmdk_local",
+    region: "us-east-1",
+    "workspace-id": "non-existent",
+    dir: fakeTmdkDir,
+  } as Arguments<Options>;
+  await expect(handler(argv2)).rejects.toThrow(ResourceNotFoundException);
+});
+
+test("deploys nothing when given an empty tmdk project", async () => {
+  fs.existsSync = jest.fn(() => true );
+  fs.readFileSync = jest.fn((path: [string]) => {
+    if (path.includes("tmdk")) {
+      return JSON.stringify(emptyTmdk, null, 4);
+    } else {
+      // covers entities.json
+      return JSON.stringify([], null, 4);
+    }
+  });
+  twinmakerMock.on(GetWorkspaceCommand).resolves({});
+
+  const argv2 = {
+    _: ["init"],
+    $0: "tmdk_local",
+    region: "us-east-1",
+    "workspace-id": workspaceId,
+    dir: fakeTmdkDir,
+  } as Arguments<Options>;
+  expect(await handler(argv2)).toBe(0);
+  expect(twinmakerMock.commandCalls(CreateComponentTypeCommand).length).toBe(
+    0
+  );
+  expect(twinmakerMock.commandCalls(CreateEntityCommand).length).toBe(0);
+  expect(twinmakerMock.commandCalls(CreateSceneCommand).length).toBe(0);
+});
+
+test("deploys successfully when given a tmdk project with one component type", async () => {
```

Comment:
Does this duplicate coverage from the integ test?

If so, would remove + de-prioritize happy-case coverage in unit tests and focus on negative cases. Mainly because I'd say the overall project is still in an alpha state so fine-grained happy cases may just cause us to have to update more places if things in the codebase change

Reply:
This does duplicate case from the functional test but I feel these tests are better for narrowing down specific cases where code fails. Functional test is more to verify that tmdk interacts with aws correctly. Also if code changes change the underlying use of aws api calls I think it is reasonable to expect developer to update the test cases as well.

Say for instance we change how deploy verifies whether a component type was created by using list component types after trying to create all of them instead of verifying each one in turn. While it will be additional work to change the test case to match this new behavior, without this fine grained test there is greater chance developer misses some edge case with their refactor.